sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobAsyncClientBase.java

```diff
@@ -545,27 +546,43 @@ public boolean isSnapshot() {
+                                                                     boolean rangeGetContentMD5) {
+        return withContext(context -> downloadToFileWithResponse(filePath, range, blockSize, options, accessConditions,
             rangeGetContentMD5, context));
     }
 
-    Mono<Void> downloadToFile(String filePath, BlobRange range, Integer blockSize, ReliableDownloadOptions options,
-        BlobAccessConditions accessConditions, boolean rangeGetContentMD5, Context context) {
+    Mono<Response<BlobProperties>> downloadToFileWithResponse(String filePath, BlobRange range, Integer blockSize,
+                                                              ReliableDownloadOptions options,
+                                                              BlobAccessConditions accessConditions,
+                                                              boolean rangeGetContentMD5, Context context) {
         if (blockSize != null) {
             Utility.assertInBounds("blockSize", blockSize, 0, BLOB_MAX_DOWNLOAD_BLOCK_SIZE);
         }
 
         return Mono.using(() -> downloadToFileResourceSupplier(filePath),
-            channel -> Mono.justOrEmpty(range)
-                .switchIfEmpty(getFullBlobRange(accessConditions))
-                .flatMapMany(rg -> Flux.fromIterable(sliceBlobRange(rg, blockSize)))
-                .flatMap(chunk -> this.download(chunk, accessConditions, rangeGetContentMD5, context)
-                    .subscribeOn(Schedulers.elastic())
-                    .flatMap(dar -> FluxUtil.writeFile(dar.body(options), channel,
-                        chunk.getOffset() - (range == null ? 0 : range.getOffset()))))
-                .then(), this::downloadToFileCleanup);
+            channel -> getPropertiesWithResponse(accessConditions).flatMap(response -> processInRange(channel, response,
+                range, blockSize, options, accessConditions, rangeGetContentMD5, context)),
+            this::downloadToFileCleanup);
+
+    }
+
+    private Mono<Response<BlobProperties>> processInRange(final AsynchronousFileChannel channel,
+                                                          final Response<BlobProperties> blobPropertiesResponse,
+                                                          BlobRange range, Integer blockSize,
+                                                          ReliableDownloadOptions options,
+                                                          BlobAccessConditions accessConditions,
+                                                          boolean rangeGetContentMD5, Context context) {
+        return Mono.justOrEmpty(range).switchIfEmpty(Mono.just(new BlobRange(0,
+            blobPropertiesResponse.getValue().getBlobSize()))).flatMapMany(rg ->
+            Flux.fromIterable(sliceBlobRange(rg, blockSize)))
+            .flatMap(chunk -> this.download(chunk, accessConditions, rangeGetContentMD5, context)
+                .subscribeOn(Schedulers.elastic())
```

Comment:
Why are you explicitly subscribing on the elastic scheduler? why not parallel? Are there an indefinite blocking tasks in the download that can grow unbounded?
https://projectreactor.io/docs/core/release/api/reactor/core/scheduler/Schedulers.html

Reply:
This is done by @jianghao. I did not have the full background on this.

Reply:
Would be nice to ger an answer on this? From the documentation, it appears boundedElastic() is preferred over elastic(), anyways.

https://projectreactor.io/docs/core/release/reference/index.html#schedulers